### ISSUE TITLE
[Common Lisp] Implement new Concept Exercise: function-definition

### ISSUE THREAD

**Issue by TheLostLambda** [exercism/v3#953]
This issue describes how to implement the `functions` concept exercise for the Common Lisp track.

## Getting started

**Please please please read the docs before starting.** Posting PRs without reading these docs will be a lot more frustrating for you during the review cycle, and exhaust Exercism's maintainers' time. So, before diving into the implementation, please read up on the following documents:

- [The features of v3](https://github.com/exercism/v3/blob/master/docs/concept-exercises.md).
- [Rationale for v3](https://github.com/exercism/v3/blob/master/docs/rationale-for-v3.md).
- [What are concept exercise and how they are structured?](https://github.com/exercism/v3/blob/master/docs/features-of-v3.md)

Please also watch the following video:

- [The Anatomy of a Concept Exercise](https://www.youtube.com/watch?v=gkbBqd7hPrA).

## Goal

This exercise should focus on teaching `defun` and its most commonly used options. The student should be comfortable defining top-level functions and using optional and keyword arguments. They should also be introduced to the usage of docstrings.

## Learning objectives

- Know how to define a global top-level function using `defun`
- Know how to annotate a `defun` with a docstring
- Be able to define a function taking parameters
- Know how to take optional arguments using `&optional`
- Know how to use `&rest` to take an arbitrary number of arguments
- Know how to take keyword arguments using `&key`
- Know how to set default values for optional / keyword arguments
- Be able to use `supplied-p` parameters with optional / keyword arguments

## Out of scope

- Anonymous functions using `lambda`
- Local functions using `flet`, `labels`, etc.
- Recursion / self-referential functions
- The somewhat arcane `&aux` keyword
- The `&allow-other-keys` keyword
- Multiple value returns with `values`
- Early returns with `return-from`
- Methods / generic functions
- Higher-order functions
- Macros

## Concepts

- `function-definition`

## Prerequisites

- `expressions`
- `integers`
- `arithmetic`
- `conditionals`

## Resources to refer to

- [Practical Common Lisp](http://www.gigamonkeys.com/book/functions.html)
- [Common Lisp Hyperspec](http://clhs.lisp.se/Body/m_defun.htm)

### Hints

- [Common Lisp Cookbook](https://lispcookbook.github.io/cl-cookbook/functions.html)

### After

Feel free to mention more niche or advanced lambda-list parameters here, things like `&allow-other-keys` or `&aux`.

## Representer

- Currently, no extra logic needs to be added to the [representer](https://github.com/exercism/common-lisp-representer).

## Analyzer

- Currently, no extra logic needs to be added to the [analyzer](https://github.com/exercism/common-lisp-analyzer).

## Implementing

Contrary to most exercises, the stub file for this exercise should not contain any `defun`'s. Comments should be used to indicate which functions the student needs to implement.

## Help

If you have any questions while implementing the exercise, please post the questions as comments in this issue.

**Comment by verdammelt** [comment]
I think this looks good. Plenty of stuff in scope that will help the student have lots of tools in their toolbox for later use.

**Comment by verdammelt** [comment]
Taking a moment to re-review this. I'm assuming that the `supplied-p-parameter` for `&optional` and `&key` are in scope? How about `&allow-other-keys`? (I'm thinking the former is in scope by the later is out of scope).

As for the in-scope 'taking function parameters' - any ideas of how we'd do that in an exericse? I'm thinking `apply` and `funcall` might be not for this exercise. So perhaps the student must write a function that takes a function argument that is when passed to another function - such as `map` or `count` etc.?

**Comment by TheLostLambda** [comment]
Thanks for taking another look at this! Yes, I think the supplied predicates would be good to include, I'll add that to in scope. As for the `&allow-other-keys`, maybe it's not a focus of the exercise, but is something to mention in `after.md`?

And "taking function parameters" was poorly worded on my part, I just meant accepting parameters in a function. I think that higher-order functions are probably out of scope here. I think they would fix better with `lambda`s in a later exercise.

**Comment by verdammelt** [comment]
In reviewing existing concepts I think that this exercise could cover the 'function', 'default_argument', 'named_parameter', and 'rest_parameter' concepts rather than 'function-definition'.

Thoughts @TheLostLambda

**Comment by TheLostLambda** [comment]
I think that sounds like a good idea! Are those concepts some of the ones shared between tracks?
